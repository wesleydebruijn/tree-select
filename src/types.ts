export interface TreeNode {
  id: string | number;
  name: string;
  children?: TreeNode[];
}

export interface TreeSettings {
  multiple: boolean;
  searchable: boolean;
  clearable: boolean;
  collapsable: boolean;
  checkable: boolean;
  openLevel: number;
  checkboxLevel: number;
  placeholder: string;
  delimiter: string;
  data?: TreeNode[];
}
