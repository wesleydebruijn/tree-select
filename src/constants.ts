import type { TreeSettings } from './types';

export const classNames = {
  wrapper: 'tree-select-wrapper',
  search: 'tree-select-search',
  dropdown: 'tree-select-dropdown',
  list: 'tree-select-list',
  item: 'tree-select-item',
  itemChildren: 'tree-select-item-children',
  itemCheckbox: 'tree-select-item-checkbox',
  itemCollapse: 'tree-select-item-collapse',
};

export const defaults: TreeSettings = {
  open: false,
  multiple: false,
  placeholder: 'Search...',
  collapsedContent: '▶',
  expandedContent: '▼',
  delimiter: ',',
};
