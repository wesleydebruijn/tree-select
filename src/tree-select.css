.tree-select-wrapper {
  position: relative;
}

.tree-select-control {
  cursor: pointer;
  border-radius: 0.5rem;
  border: 1px solid #e5e7eb;
  padding: 0.5rem  0.75rem;
  font-size: 0.875rem;
  line-height: 1.25rem;
}

.tree-select-control.focus {
  border-color: #3b82f6;
  outline: none;
  box-shadow: 0 0 0 3px rgba(59, 130, 246, 0.5);
}

.tree-select-heading {
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.tree-select-search {
  width: 100%;
  outline: none;
  display: block;
  width: 100%;
  border-radius: 0.5rem;
  border: 1px solid #e5e7eb;
  padding: 0.25rem 0.5rem;
  font-size: 0.75rem;
}

.tree-select-search::placeholder {
  color: #6b7280;
}

.tree-select-search:focus {
  border-color: #3b82f6;
  outline: none;
  box-shadow: 0 0 0 3px rgba(59, 130, 246, 0.5);
}

.tree-select-dropdown {
  position: absolute;
  display: flex;
  flex-direction: column;
  gap: 5px;
  top: calc(100% + 5px);
  left: 0;
  background-color: #ffffff;
  max-height: 300px;
  z-index: 50;
  padding: 10px;
  border-radius: 0.5rem;
  border: 1px solid #e5e7eb;
  box-shadow: 0 4px 6px -1px rgba(0, 0, 0, 0.1), 0 2px 4px -1px rgba(0, 0, 0, 0.06);
}

.tree-select-wrapper:hover .tree-select-clear {
  opacity: 1;
}

.tree-select-clear {
  position: absolute;
  transition: opacity 0.2s ease-in-out;
  right: 5px;
  top: 50%;
  opacity: 0;
  padding: 0 10px;
  font-size: 0.75rem;
  transform: translateY(-50%);
  cursor: pointer;
}

.tree-select-list {
  width: 100%;
  overflow-y: auto;
  scrollbar-width: none;
  -ms-overflow-style: none;
}

.tree-select-label {
  display: flex;
  align-items: center;
  gap: 2px;
}

.tree-select-children {
  margin-left: 18px;
}

.tree-select-label {
  cursor: pointer;
}

.tree-select-collapse {
  background-image: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 16 16'%3E%3Cpath d='M8 13.1l-8-8 2.1-2.2 5.9 5.9 5.9-5.9 2.1 2.2z'/%3E%3C/svg%3E");
  height: 10px;
  width: 10px;
  transition: transform 0.1s ease-in-out;
}

.tree-select-collapse.collapsed {
  transform: rotate(-90deg);
}
