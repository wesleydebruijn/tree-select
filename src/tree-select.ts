import {
  createItems,
  itemValues,
  itemsDepth,
  searchItems,
  selectItem,
  selectItemRange,
  selectItemsByValues,
  updateItems,
} from "./utils/core";
import { className, create, debounce, visible } from "./utils/dom";
import { getInputElement, getInputValues, setInputValues } from "./utils/input";

import type { Data, TreeItem, TreeSettings } from "./types";

export class TreeSelect {
  public settings: TreeSettings = {
    open: false,
    clearable: true,
    searchable: true,
    collapsible: true,
    focus: "focus",
    delimiter: ",",
    depthCollapsible: 0,
    depthCollapsed: 0,
    depthCheckboxes: 0,
    depthValues: "last",
    text: {
      selected: "selected",
      loading: "Loading...",
      search: "Search...",
    },
    html: {},
  };

  // UI
  public opened = false;
  public loaded = false;
  public loading = false;
  public mounted = false;
  public search = "";

  // Data
  public data: Data[] = [];
  public values: string[] = [];
  public items: Map<string, TreeItem> = new Map();
  public lastItem: TreeItem | null = null;
  public depth = 0;
  public depthValues = 0;

  // DOM
  private rootElement: HTMLInputElement | HTMLSelectElement;
  private controlElement: HTMLElement | null = null;
  private wrapperElement: HTMLElement | null = null;
  private dropdownElement: HTMLElement | null = null;
  private loadingElement: HTMLElement | null = null;

  constructor(
    input: HTMLInputElement | HTMLSelectElement | string,
    settings: Partial<TreeSettings> = {},
  ) {
    this.settings = Object.assign(this.settings, settings);

    this.onLoad = this.onLoad.bind(this);
    this.onChange = this.onChange.bind(this);
    this.onFocus = this.onFocus.bind(this);
    this.onSearch = this.onSearch.bind(this);
    this.onKeyDown = this.onKeyDown.bind(this);
    this.onClear = this.onClear.bind(this);
    this.mountItem = this.mountItem.bind(this);
    this.renderItem = this.renderItem.bind(this);

    // initialize the root input element
    this.rootElement = getInputElement(input);
    this.rootElement.treeSelect = this;
    this.rootElement.addEventListener("change", this.onChange);

    // initialize the selected values
    this.values = getInputValues(this.rootElement, this.settings.delimiter);

    // mount initial HTML to the DOM, without fetching data or rendering tree
    this.initMount();

    if (this.settings.open) this.open();
  }

  public open(): void {
    if (this.opened) return;
    if (!this.loaded && !this.loading) debounce(() => this.load(), 0)();

    this.opened = true;
    className(this.controlElement, this.settings.focus, true);
    visible(this.dropdownElement, true);

    this.onOpen();
  }

  public close(): void {
    if (!this.opened) return;

    this.opened = false;
    className(this.controlElement, this.settings.focus, false);
    visible(this.dropdownElement, false);

    this.onClose();
  }

  public async load(): Promise<void> {
    if (!this.settings.dataSrc) return this.onLoad(this.settings.data || []);

    return fetch(this.settings.dataSrc)
      .then((response) => response.json())
      .then(this.onLoad);
  }

  public destroy(): void {
    document.removeEventListener("mousedown", this.onFocus, true);
    document.removeEventListener("focus", this.onFocus, true);

    if (this.wrapperElement) this.wrapperElement.remove();

    visible(this.rootElement, true);

    this.rootElement.treeSelect = null;
    this.rootElement.removeEventListener("change", this.onChange);
  }

  private initMount(): void {
    // create the wrapper element
    this.wrapperElement = create("div", "wrapper", this.settings.html);
    this.wrapperElement.addEventListener("keydown", this.onKeyDown);
    this.rootElement.after(this.wrapperElement);

    // create the controlled element
    this.controlElement = create("div", "control", this.settings.html);
    this.controlElement.tabIndex = 0;
    this.controlElement.innerHTML = `${this.values.length} ${this.settings.text.selected}`;
    this.controlElement.addEventListener("focus", this.onFocus);
    this.wrapperElement.appendChild(this.controlElement);

    if (this.settings.clearable) {
      const clearElement = create("span", "clear", this.settings.html);
      clearElement.innerHTML = "X";
      clearElement.addEventListener("click", this.onClear);
      this.wrapperElement.appendChild(clearElement);
    }

    // create the dropdown element
    this.dropdownElement = create("div", "dropdown", this.settings.html);
    visible(this.dropdownElement, false);
    this.wrapperElement.appendChild(this.dropdownElement);

    // create the search element
    if (this.settings.searchable) {
      const searchElement = create("input", "search", this.settings.html);
      searchElement.type = "search";
      if (this.settings.text.search) searchElement.placeholder = this.settings.text.search;
      searchElement.addEventListener("keyup", debounce(this.onSearch, 100));
      searchElement.addEventListener("search", debounce(this.onSearch, 100));
      this.dropdownElement.appendChild(searchElement);
    }

    // create the loading element
    this.loadingElement = create("div", "loading", this.settings.html);
    if (this.settings.text.loading) this.loadingElement.innerHTML = this.settings.text.loading;
    this.dropdownElement.appendChild(this.loadingElement);

    // add event listeners
    document.addEventListener("mousedown", this.onFocus, true);
    document.addEventListener("focus", this.onFocus, true);

    // hide the initial input element
    visible(this.rootElement, false);
  }

  private mountItems(): void {
    // create the list element
    const listElement = create("div", "list", this.settings.html);

    // create the items
    this.items = new Map<string, TreeItem>();
    createItems(this.items, this.data, this.mountItem, listElement);
    this.depth = itemsDepth(this.items);

    this.depthValues =
      this.settings.depthValues === "last" ? this.depth : this.settings.depthValues;

    selectItemsByValues(this.items, this.values, this.depthValues);

    this.render();

    // append the list element to the dropdown element
    visible(this.loadingElement, false);
    if (this.dropdownElement) this.dropdownElement.appendChild(listElement);

    this.mounted = true;
  }

  private mountItem(item: TreeItem, element: HTMLElement) {
    const isCollapsible =
      item.children && item.depth >= this.settings.depthCollapsible && this.settings.collapsible;
    const isCheckable = item.depth >= this.settings.depthCheckboxes;

    item.collapsed = item.depth >= this.settings.depthCollapsed;

    // create root item element
    item.itemElement = create("div", "item", this.settings.html);

    // create label with collapse element, checkbox and name
    const labelElement = create("div", "label", this.settings.html);

    // create collapse element
    if (isCollapsible) {
      item.collapseElement = create("div", "collapse", this.settings.html);
      labelElement.appendChild(item.collapseElement);
    }

    // create checkbox element
    if (isCheckable) {
      item.checkboxElement = create("input", "checkbox", this.settings.html);
      item.checkboxElement.type = "checkbox";
      item.checkboxElement.addEventListener("click", (event) => this.onItemSelect(event, item));
      labelElement.appendChild(item.checkboxElement);
    }

    const labelSpan = create("span", "labelSpan", this.settings.html);
    labelSpan.innerHTML = item.name;
    labelElement.appendChild(labelSpan);
    if (isCollapsible || isCheckable) {
      labelElement.addEventListener("click", (event) =>
        isCollapsible ? this.onItemCollapse(event, item) : this.onItemSelect(event, item),
      );
    }

    // append label and children element to root item element
    item.itemElement.appendChild(labelElement);
    if (item.children) {
      item.childrenElement = create("div", "children", this.settings.html);
      item.itemElement.appendChild(item.childrenElement);
    }

    // append item element to the parent element
    element.appendChild(item.itemElement);
  }

  private render(): void {
    for (const item of this.items.values()) {
      this.renderItem(item);
    }

    if (this.controlElement)
      this.controlElement.innerHTML = `${this.values.length} ${this.settings.text.selected}`;
  }

  private renderItem(item: TreeItem): void {
    visible(
      item.childrenElement,
      !item.collapsed || item.depth < this.settings.depthCollapsible || !this.settings.collapsible,
    );
    visible(item.itemElement, !item.hidden);

    className(item.collapseElement, "collapsed", item.collapsed);

    if (!item.checkboxElement) return;

    item.checkboxElement.checked = item.checked;
    item.checkboxElement.indeterminate = item.indeterminate;
  }

  private onLoad(data: Data[]): void {
    this.data = data;
    this.loaded = true;
    this.loading = false;

    if (!this.mounted) this.mountItems();
    if (this.settings.onLoad) this.settings.onLoad(data);
  }

  private onItemSelect(event: MouseEvent, item: TreeItem): void {
    event.stopPropagation();

    if (event.shiftKey && this.lastItem) {
      selectItemRange(this.items, this.lastItem, item);
    } else {
      selectItem(this.items, item);

      this.lastItem = item;
    }

    // update input values
    this.values = itemValues(this.items, this.depthValues);
    setInputValues(this.rootElement, this.values, this.settings.delimiter);

    this.render();

    if (this.settings.onSelect) this.settings.onSelect(this.values);
  }

  private onItemCollapse(_event: Event, item: TreeItem): void {
    item.collapsed = !item.collapsed;

    this.renderItem(item);
  }

  private onSearch(event: Event): void {
    const search = (event.target as HTMLInputElement).value;
    if (search === this.search) return;

    this.search = search;
    searchItems(this.items, search);

    this.render();

    if (this.settings.onSearch) this.settings.onSearch(search);
  }

  private onKeyDown(event: KeyboardEvent): void {
    if (event.key === "Escape") this.close();
  }

  private onClear(): void {
    updateItems(this.items, { checked: false, indeterminate: false });

    // update input values
    this.values = [];
    setInputValues(this.rootElement, this.values, this.settings.delimiter);

    this.render();

    if (this.settings.onClear) this.settings.onClear();
  }

  private onOpen(): void {
    if (this.settings.onOpen) this.settings.onOpen();
  }

  private onClose(): void {
    if (this.settings.onClose) this.settings.onClose();
  }

  private onFocus(event: Event): void {
    if (this.wrapperElement?.contains(event.target as HTMLElement)) {
      this.open();
    } else {
      this.close();
    }
  }

  private onChange(_event: Event): void {
    this.values = getInputValues(this.rootElement, this.settings.delimiter);
    if (this.mounted) selectItemsByValues(this.items, this.values, this.depthValues);

    this.render();
  }
}

export type { TreeItem, TreeSettings, Data };
