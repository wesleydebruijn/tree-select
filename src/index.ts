import { defaults, classNames } from './constants';
import {
  createDiv,
  createInput,
  createLabel,
  createSpan,
  createCheckbox,
  visible,
} from './utils/dom';
import { debounce } from './utils/global';
import { getInputElement, getInputValues } from './utils/input';
import { stripId, update, updateChildren, updateAncestors } from './utils/items';

import type { TreeItem, TreeRecord, TreeSettings } from './types';

declare global {
  interface HTMLElement {
    treeSelect: TreeSelect | null;
  }
}

export class TreeSelect {
  public settings: TreeSettings;

  public opened: boolean = false;
  public loaded: boolean = false;
  public loading: boolean = false;
  public search: string = '';

  public data: TreeRecord[] = [];
  public selected: string[] = [];

  public items: Map<string, TreeItem> = new Map();
  public itemLevels: number = 0;

  public inputElement: HTMLInputElement | HTMLSelectElement;
  public wrapperElement: HTMLElement | null = null;
  public searchElement: HTMLInputElement | null = null;
  public dropdownElement: HTMLElement | null = null;
  public listElement: HTMLElement | null = null;

  constructor(
    input: HTMLInputElement | HTMLSelectElement | string,
    settings: Partial<TreeSettings> = {}
  ) {
    this.settings = { ...defaults, ...settings };
    this.onFocus = this.onFocus.bind(this);
    this.onSearch = this.onSearch.bind(this);

    // initialize the root input element
    this.inputElement = getInputElement(input);

    if (this.inputElement.treeSelect) {
      throw new Error('TreeSelect already initialized on this element');
    }

    this.inputElement.treeSelect = this;

    // initialize the selected values from the input element
    this.selected = getInputValues(this.inputElement, this.settings.delimiter);

    this.data = this.settings.data || [];

    this.mount();

    if (this.settings.open) this.open();
  }

  public mount(): void {
    // create the wrapper element
    this.wrapperElement = createDiv(classNames.wrapper, this.settings.wrapperClassName);
    this.inputElement.after(this.wrapperElement);

    // create the search element
    this.searchElement = createInput(
      classNames.search,
      this.settings.searchClassName || this.inputElement.className
    );
    this.searchElement.placeholder = this.settings.placeholder;
    this.searchElement.addEventListener('keyup', debounce(this.onSearch, 100));
    this.wrapperElement.appendChild(this.searchElement);

    // create the dropdown element
    this.dropdownElement = createDiv(classNames.dropdown, this.settings.dropdownClassName);
    visible(this.dropdownElement, false);
    this.wrapperElement.appendChild(this.dropdownElement);

    // add event listeners
    document.addEventListener('mousedown', this.onFocus, true);
    document.addEventListener('focus', this.onFocus, true);

    // hide the initial input element
    // visible(this.inputElement, false);
  }

  public open(): void {
    if (this.opened) return;
    if (!this.loaded && !this.loading) this.load();

    this.opened = true;
    visible(this.dropdownElement, true);

    this.onOpen();
  }

  public close(): void {
    if (!this.opened) return;

    this.opened = false;
    visible(this.dropdownElement, false);

    this.onClose();
  }

  public async load(): Promise<void> {
    if (!this.settings.src) return this.onLoad();

    return fetch(this.settings.src)
      .then(response => response.json())
      .then((data: TreeRecord[]) => {
        this.data = data;
        this.onLoad();
      });
  }

  public destroy(): void {
    document.removeEventListener('mousedown', this.onFocus, true);
    document.removeEventListener('focus', this.onFocus, true);

    this.wrapperElement?.remove();
    visible(this.inputElement, true);

    this.inputElement.treeSelect = null;
  }

  private createList(): void {
    this.listElement?.remove();

    this.listElement = createDiv(classNames.list, this.settings.listClassName);
    this.createItems(this.data, 0, this.listElement);
    this.selectItems(this.selected);

    this.updateDOM();

    if (this.dropdownElement) this.dropdownElement.appendChild(this.listElement);
  }

  private createItems(
    data: TreeRecord[],
    level: number = 0,
    parentElement?: HTMLElement | null,
    parent?: string
  ): void {
    if (level > this.itemLevels) this.itemLevels = level;

    data.forEach(record => {
      let item: TreeItem = {
        id: `${level}-${record.id}`,
        parent,
        children: record.children?.map(child => `${level + 1}-${child.id}`),
        level,
        name: record.name,
        checked: false,
        indeterminate: false,
        collapsed: true,
        hidden: false,
        itemElement: null,
        checkboxElement: null,
        collapseElement: null,
        childrenElement: null,
      };

      const itemElement = createDiv(classNames.item, this.settings.itemClassName);
      const labelElement = createLabel(classNames.itemLabel);
      itemElement.appendChild(labelElement);

      const checkboxElement = createCheckbox(
        classNames.itemCheckbox,
        this.settings.checkboxClassName
      );
      checkboxElement.addEventListener('click', event => this.onItemSelect(event, item));

      let collapseElement: HTMLElement | null = null;
      let childrenElement: HTMLElement | null = null;
      if (record.children) {
        collapseElement = createDiv(classNames.itemCollapse, this.settings.collapseClassName);
        collapseElement.addEventListener('click', event => this.onItemCollapse(event, item));
        labelElement.appendChild(collapseElement);

        childrenElement = createDiv(classNames.itemChildren);
        itemElement.appendChild(childrenElement);

        this.createItems(record.children, level + 1, childrenElement, item.id);
      }

      labelElement.appendChild(checkboxElement);
      labelElement.appendChild(createSpan(record.name));

      if (parentElement) parentElement.appendChild(itemElement);

      item.itemElement = itemElement;
      item.checkboxElement = checkboxElement;
      item.collapseElement = collapseElement;
      item.childrenElement = childrenElement;

      this.items.set(item.id, item);
    });
  }

  private selectItems(selected: string[]): void {
    let ancestors = new Set<string>();

    update(this.items, item => {
      const checked = item.level === this.itemLevels && selected.includes(stripId(item.id));

      if (checked && item.parent) ancestors.add(item.parent);

      return { checked, indeterminate: false };
    });

    ancestors.forEach(id => {
      const item = this.items.get(id)!;

      this.propagateItems(item);
      updateAncestors(this.items, item, item => this.propagateItems(item));
    });
  }

  private propagateItems(item: TreeItem): void {
    if (!item.children) return;

    const children = item.children.map(id => this.items.get(id)!);
    const allChecked = children.every(child => child.checked);
    const someChecked = children.some(child => child.checked || child.indeterminate);

    item.checked = allChecked;
    item.indeterminate = !allChecked && someChecked;
  }

  private updateDOM(): void {
    this.items.forEach(item => {
      visible(item.childrenElement, !item.collapsed);
      visible(item.itemElement, !item.hidden);

      if (item.collapseElement)
        item.collapseElement.innerHTML = item.collapsed
          ? this.settings.collapsedContent
          : this.settings.expandedContent;

      if (item.checkboxElement) {
        item.checkboxElement.checked = item.checked;
        item.checkboxElement.indeterminate = item.indeterminate;
      }
    });
  }

  // callbacks

  private onLoad(): void {
    this.loaded = true;
    this.loading = false;

    this.createList();

    if (this.settings.onLoad) this.settings.onLoad(this.data);
  }

  private onItemSelect(_event: Event, item: TreeItem): void {
    item.checked = !item.checked;
    item.indeterminate = false;

    updateChildren(this.items, item, { checked: item.checked, indeterminate: false });
    updateAncestors(this.items, item, item => this.propagateItems(item));

    this.updateDOM();

    if (this.settings.onSelect) this.settings.onSelect(this.selected);
  }

  private onItemCollapse(event: Event, item: TreeItem): void {
    event.preventDefault();

    item.collapsed = !item.collapsed;

    this.updateDOM();
  }

  private onSearch(event: Event): void {
    const search = (event.target as HTMLInputElement).value;
    if (search === this.search) return;

    this.search = search;

    if (this.search.length === 0) {
      update(this.items, { hidden: false, collapsed: true });
    } else {
      update(this.items, { hidden: true, collapsed: false });
      update(this.items, item => {
        const match = item.name.toLowerCase().includes(this.search.toLowerCase());
        if (!match) return;

        item.hidden = false;

        updateChildren(this.items, item, { hidden: false });
        updateAncestors(this.items, item, { hidden: false });
      });
    }

    this.updateDOM();

    if (this.settings.onSearch) this.settings.onSearch(search);
  }

  private onOpen(): void {
    if (this.settings.onOpen) this.settings.onOpen();
  }

  private onClose(): void {
    if (this.settings.onClose) this.settings.onClose();
  }

  private onFocus(event: Event): void {
    if (this.wrapperElement?.contains(event.target as HTMLElement)) {
      this.open();
    } else {
      this.close();
    }
  }
}

export { TreeItem, TreeRecord, TreeSettings };
