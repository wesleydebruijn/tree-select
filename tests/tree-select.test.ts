import { TreeSelect } from "src/tree-select";

describe("TreeSelect", () => {
  beforeEach(() => {
    document.body.innerHTML = "";
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  it("should initialize with default settings", () => {
    const input = document.createElement("input");
    document.body.appendChild(input);
    const treeSelect = new TreeSelect(input);

    expect(treeSelect.settings).toEqual({
      open: false,
      clearable: true,
      searchable: true,
      collapsible: true,
      results: false,
      mode: "vertical",
      delimiter: ",",
      depthCollapsible: 0,
      depthCollapsed: 0,
      depthCheckboxes: 0,
      depthValues: "last",
      focus: "focus",
      disabled: "disabled",
      html: {},
      text: {
        selected: "selected",
        loading: "Loading...",
        search: "Search...",
      },
    });

    expect(document.body.innerHTML).toMatchSnapshot();
  });

  it("should initialize with custom settings", () => {
    const input = document.createElement("input");
    document.body.appendChild(input);
    const treeSelect = new TreeSelect(input, {
      open: true,
      clearable: false,
      searchable: false,
      collapsible: false,
      results: true,
      mode: "vertical",
      delimiter: "|",
      depthCollapsible: 2,
      depthCollapsed: 1,
      depthCheckboxes: 3,
      depthValues: 2,
      text: {
        selected: "items selected",
        loading: "please wait...",
        search: "filter...",
      },
      html: {
        wrapper: { className: "custom-wrapper" },
        control: { className: "custom-control" },
      },
    });

    expect(treeSelect.settings).toEqual({
      open: true,
      clearable: false,
      searchable: false,
      collapsible: false,
      results: true,
      mode: "vertical",
      delimiter: "|",
      depthCollapsible: 2,
      depthCollapsed: 1,
      depthCheckboxes: 3,
      depthValues: 2,
      focus: "focus",
      disabled: "disabled",
      text: {
        selected: "items selected",
        loading: "please wait...",
        search: "filter...",
      },
      html: {
        wrapper: { className: "custom-wrapper" },
        control: { className: "custom-control" },
      },
    });

    expect(document.body.innerHTML).toMatchSnapshot();
  });

  it("should load data from settings", async () => {
    const input = document.createElement("input");
    document.body.appendChild(input);
    const treeSelect = new TreeSelect(input, {
      data: [
        {
          id: "1",
          name: "Parent",
          children: [
            { id: "2", name: "Child 1" },
            { id: "3", name: "Child 2" },
          ],
        },
      ],
    });

    await treeSelect.load();
    expect(document.body.innerHTML).toMatchSnapshot();
  });

  it("should load data from URL", async () => {
    const input = document.createElement("input");
    document.body.appendChild(input);
    const treeSelect = new TreeSelect(input, {
      dataSrc: "/api/data",
    });

    global.fetch = jest.fn().mockImplementation(() =>
      Promise.resolve({
        json: () =>
          Promise.resolve([
            {
              id: "1",
              name: "Parent",
              children: [
                { id: "2", name: "Child 1" },
                { id: "3", name: "Child 2" },
              ],
            },
          ]),
      })
    );

    await treeSelect.load();
    expect(document.body.innerHTML).toMatchSnapshot();
    expect(global.fetch).toHaveBeenCalledWith("/api/data");
  });

  it("should handle open and close", () => {
    const input = document.createElement("input");
    document.body.appendChild(input);
    const treeSelect = new TreeSelect(input);

    treeSelect.open();
    expect(document.body.innerHTML).toMatchSnapshot();

    treeSelect.close();
    expect(document.body.innerHTML).toMatchSnapshot();
  });

  it("should handle search", () => {
    const input = document.createElement("input");
    document.body.appendChild(input);
    const treeSelect = new TreeSelect(input, {
      data: [
        {
          id: "1",
          name: "Parent",
          children: [
            { id: "2", name: "Child 1" },
            { id: "3", name: "Child 2" },
          ],
        },
      ],
    });

    treeSelect.open();
    const searchInput = document.querySelector(
      ".tree-select-search"
    ) as HTMLInputElement;
    searchInput.value = "Child 1";
    searchInput.dispatchEvent(new Event("input"));

    expect(document.body.innerHTML).toMatchSnapshot();
  });

  it("should handle destroy", () => {
    const input = document.createElement("input");
    document.body.appendChild(input);
    const treeSelect = new TreeSelect(input);

    treeSelect.destroy();
    expect(document.body.innerHTML).toBe('<input class="" style="">');
  });

  it("should clear selection when clear button is clicked", () => {
    const input = document.createElement("input");
    document.body.appendChild(input);
    const treeSelect = new TreeSelect(input, {
      data: [
        {
          id: "1",
          name: "Parent",
          children: [
            { id: "2", name: "Child 1" },
            { id: "3", name: "Child 2" },
          ],
        },
      ],
    });

    treeSelect.open();
    // Simulate selecting an item
    input.value = "2";
    input.dispatchEvent(new Event("change"));

    // Click the clear button
    const clearButton = document.querySelector(
      ".tree-select-clear"
    ) as HTMLElement;
    clearButton.click();

    expect(input.value).toBe("");
    expect(document.body.innerHTML).toMatchSnapshot();
  });

  it("should handle disable and enable", () => {
    const input = document.createElement("input");
    document.body.appendChild(input);
    const treeSelect = new TreeSelect(input);

    // Test disable
    treeSelect.disable();
    expect(input.disabled).toBe(true);
    expect(document.body.innerHTML).toMatchSnapshot();

    // Test enable
    treeSelect.enable();
    expect(input.disabled).toBe(false);
    expect(document.body.innerHTML).toMatchSnapshot();
  });
});
