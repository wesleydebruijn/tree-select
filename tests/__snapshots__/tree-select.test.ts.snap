// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeSelect should clear selection when clear button is clicked 1`] = `"<input class="treeselected" style="display: none;"><div class="tree-select-wrapper focus"><div class="tree-select-control" tabindex="0">0 selected</div><span class="tree-select-clear">X</span><div class="tree-select-dropdown" style=""><input class="tree-select-search" type="search" placeholder="Search..."><div class="tree-select-loading">Loading...</div></div></div>"`;

exports[`TreeSelect should handle disable and enable 1`] = `"<input class="treeselected" style="display: none;" disabled=""><div class="tree-select-wrapper disabled"><div class="tree-select-control disabled" tabindex="0">0 selected</div><span class="tree-select-clear">X</span><div class="tree-select-dropdown" style="display: none;"><input class="tree-select-search" type="search" placeholder="Search..."><div class="tree-select-loading">Loading...</div></div></div>"`;

exports[`TreeSelect should handle disable and enable 2`] = `"<input class="treeselected" style="display: none;"><div class="tree-select-wrapper"><div class="tree-select-control" tabindex="0">0 selected</div><span class="tree-select-clear">X</span><div class="tree-select-dropdown" style="display: none;"><input class="tree-select-search" type="search" placeholder="Search..."><div class="tree-select-loading">Loading...</div></div></div>"`;

exports[`TreeSelect should handle open and close 1`] = `"<input class="treeselected" style="display: none;"><div class="tree-select-wrapper focus"><div class="tree-select-control" tabindex="0">0 selected</div><span class="tree-select-clear">X</span><div class="tree-select-dropdown" style=""><input class="tree-select-search" type="search" placeholder="Search..."><div class="tree-select-loading">Loading...</div></div></div>"`;

exports[`TreeSelect should handle open and close 2`] = `"<input class="treeselected" style="display: none;"><div class="tree-select-wrapper"><div class="tree-select-control" tabindex="0">0 selected</div><span class="tree-select-clear">X</span><div class="tree-select-dropdown" style="display: none;"><input class="tree-select-search" type="search" placeholder="Search..."><div class="tree-select-loading">Loading...</div></div></div>"`;

exports[`TreeSelect should handle search 1`] = `"<input class="treeselected" style="display: none;"><div class="tree-select-wrapper focus"><div class="tree-select-control" tabindex="0">0 selected</div><span class="tree-select-clear">X</span><div class="tree-select-dropdown" style=""><input class="tree-select-search" type="search" placeholder="Search..."><div class="tree-select-loading">Loading...</div></div></div>"`;

exports[`TreeSelect should initialize with custom settings 1`] = `"<input class="treeselected" style="display: none;"><div class="tree-select-wrapper custom-wrapper focus"><div class="tree-select-control custom-control" tabindex="0">0 items selected</div><div class="tree-select-dropdown" style=""><div class="tree-select-loading">please wait...</div></div></div>"`;

exports[`TreeSelect should initialize with default settings 1`] = `"<input class="treeselected" style="display: none;"><div class="tree-select-wrapper"><div class="tree-select-control" tabindex="0">0 selected</div><span class="tree-select-clear">X</span><div class="tree-select-dropdown" style="display: none;"><input class="tree-select-search" type="search" placeholder="Search..."><div class="tree-select-loading">Loading...</div></div></div>"`;

exports[`TreeSelect should load data from URL 1`] = `"<input class="treeselected" style="display: none;"><div class="tree-select-wrapper"><div class="tree-select-control" tabindex="0">0 selected</div><span class="tree-select-clear">X</span><div class="tree-select-dropdown" style="display: none;"><input class="tree-select-search" type="search" placeholder="Search..."><div class="tree-select-loading" style="display: none;">Loading...</div><div class="tree-select-list-container"><div class="tree-select-list"><div class="tree-select-item"><div class="tree-select-label"><div class="tree-select-collapse collapsed"></div><input class="tree-select-checkbox" type="checkbox"><span class="tree-select-label-span">Parent</span></div><div class="tree-select-children" style="display: none;"><div class="tree-select-item"><div class="tree-select-label"><input class="tree-select-checkbox" type="checkbox"><span class="tree-select-label-span">Child 1</span></div></div><div class="tree-select-item"><div class="tree-select-label"><input class="tree-select-checkbox" type="checkbox"><span class="tree-select-label-span">Child 2</span></div></div></div></div></div></div></div></div>"`;

exports[`TreeSelect should load data from settings 1`] = `"<input class="treeselected" style="display: none;"><div class="tree-select-wrapper"><div class="tree-select-control" tabindex="0">0 selected</div><span class="tree-select-clear">X</span><div class="tree-select-dropdown" style="display: none;"><input class="tree-select-search" type="search" placeholder="Search..."><div class="tree-select-loading" style="display: none;">Loading...</div><div class="tree-select-list-container"><div class="tree-select-list"><div class="tree-select-item"><div class="tree-select-label"><div class="tree-select-collapse collapsed"></div><input class="tree-select-checkbox" type="checkbox"><span class="tree-select-label-span">Parent</span></div><div class="tree-select-children" style="display: none;"><div class="tree-select-item"><div class="tree-select-label"><input class="tree-select-checkbox" type="checkbox"><span class="tree-select-label-span">Child 1</span></div></div><div class="tree-select-item"><div class="tree-select-label"><input class="tree-select-checkbox" type="checkbox"><span class="tree-select-label-span">Child 2</span></div></div></div></div></div></div></div></div>"`;
